File: packages/firebase_database/ios/Classes/FirebaseDatabasePlugin.m
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "FirebaseDatabasePlugin.h"

#import <Firebase/Firebase.h>

static FlutterError *getFlutterError(NSError *error) {
  return [FlutterError errorWithCode:[NSString stringWithFormat:@"Error %d", (int)error.code]
                             message:error.domain
                             details:error.localizedDescription];
}

static NSDictionary *getDictionaryFromError(NSError *error) {
  return @{
    @"code" : @(error.code),
    @"message" : error.domain ?: [NSNull null],
    @"details" : error.localizedDescription ?: [NSNull null],
  };
}

@interface FLTFirebaseDatabasePlugin ()

@end

FIRDatabaseReference *getReference(FIRDatabase *database, NSDictionary *arguments) {
  NSString *path = arguments[@"path"];
  FIRDatabaseReference *ref = database.reference;
  if ([path length] > 0) ref = [ref child:path];
  return ref;
}

FIRDatabaseQuery *getDatabaseQuery(FIRDatabase *database, NSDictionary *arguments) {
  FIRDatabaseQuery *query = getReference(database, arguments);
  NSDictionary *parameters = arguments[@"parameters"];
  NSString *orderBy = parameters[@"orderBy"];
  if ([orderBy isEqualToString:@"child"]) {
    query = [query queryOrderedByChild:parameters[@"orderByChildKey"]];
  } else if ([orderBy isEqualToString:@"key"]) {
    query = [query queryOrderedByKey];
  } else if ([orderBy isEqualToString:@"value"]) {
    query = [query queryOrderedByValue];
  } else if ([orderBy isEqualToString:@"priority"]) {
    query = [query queryOrderedByPriority];
  }
  id startAt = parameters[@"startAt"];
  if (startAt) {
    id startAtKey = parameters[@"startAtKey"];
    if (startAtKey) {
      query = [query queryStartingAtValue:startAt childKey:startAtKey];
    } else {
      query = [query queryStartingAtValue:startAt];
    }
  }
  id endAt = parameters[@"endAt"];
  if (endAt) {
    id endAtKey = parameters[@"endAtKey"];
    if (endAtKey) {
      query = [query queryEndingAtValue:endAt childKey:endAtKey];
    } else {
      query = [query queryEndingAtValue:endAt];
    }
  }
  id equalTo = parameters[@"equalTo"];
  if (equalTo) {
    id equalToKey = parameters[@"equalToKey"];
    if (equalToKey) {
      query = [query queryEqualToValue:equalTo childKey:equalToKey];
    } else {
      query = [query queryEqualToValue:equalTo];
    }
  }
  NSNumber *limitToFirst = parameters[@"limitToFirst"];
  if (limitToFirst) {
    query = [query queryLimitedToFirst:limitToFirst.intValue];
  }
  NSNumber *limitToLast = parameters[@"limitToLast"];
  if (limitToLast) {
    query = [query queryLimitedToLast:limitToLast.intValue];
  }
  return query;
}

FIRDataEventType parseEventType(NSString *eventTypeString) {
  if ([@"_EventType.childAdded" isEqual:eventTypeString]) {
    return FIRDataEventTypeChildAdded;
  } else if ([@"_EventType.childRemoved" isEqual:eventTypeString]) {
    return FIRDataEventTypeChildRemoved;
  } else if ([@"_EventType.childChanged" isEqual:eventTypeString]) {
    return FIRDataEventTypeChildChanged;
  } else if ([@"_EventType.childMoved" isEqual:eventTypeString]) {
    return FIRDataEventTypeChildMoved;
  } else if ([@"_EventType.value" isEqual:eventTypeString]) {
    return FIRDataEventTypeValue;
  }
  assert(false);
  return 0;
}

id roundDoubles(id value) {
  // Workaround for https://github.com/firebase/firebase-ios-sdk/issues/91
  // The Firebase iOS SDK sometimes returns doubles when ints were stored.
  // We detect doubles that can be converted to ints without loss of precision
  // and convert them.
  if ([value isKindOfClass:[NSNumber class]]) {
    CFNumberType type = CFNumberGetType((CFNumberRef)value);
    if (type == kCFNumberDoubleType || type == kCFNumberFloatType) {
      if ((double)(long long)[value doubleValue] == [value doubleValue]) {
        return [NSNumber numberWithLongLong:(long long)[value doubleValue]];
      }
    }
  } else if ([value isKindOfClass:[NSArray class]]) {
    NSMutableArray *result = [NSMutableArray arrayWithCapacity:[value count]];
    [value enumerateObjectsUsingBlock:^(id obj, NSUInteger idx, BOOL *stop) {
      [result addObject:roundDoubles(obj)];
    }];
    return result;
  } else if ([value isKindOfClass:[NSDictionary class]]) {
    NSMutableDictionary *result = [NSMutableDictionary dictionaryWithCapacity:[value count]];
    [value enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *stop) {
      result[key] = roundDoubles(obj);
    }];
    return result;
  }
  return value;
}

@interface FLTFirebaseDatabasePlugin ()
@property(nonatomic, retain) FlutterMethodChannel *channel;
@end

@implementation FLTFirebaseDatabasePlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  FlutterMethodChannel *channel =
      [FlutterMethodChannel methodChannelWithName:@"plugins.flutter.io/firebase_database"
                                  binaryMessenger:[registrar messenger]];
  FLTFirebaseDatabasePlugin *instance = [[FLTFirebaseDatabasePlugin alloc] init];
  instance.channel = channel;
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (instancetype)init {
  self = [super init];
  if (self) {
    if (![FIRApp defaultApp]) {
      [FIRApp configure];
    }
    self.updatedSnapshots = [NSMutableDictionary new];
  }
  return self;
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
  FIRDatabase *database;
  NSString *appName = call.arguments[@"app"];
  NSString *databaseURL = call.arguments[@"databaseURL"];
  if (![appName isEqual:[NSNull null]] && ![databaseURL isEqual:[NSNull null]]) {
    database = [FIRDatabase databaseForApp:[FIRApp appNamed:appName] URL:databaseURL];
  } else if (![appName isEqual:[NSNull null]]) {
    database = [FIRDatabase databaseForApp:[FIRApp appNamed:appName]];
  } else if (![databaseURL isEqual:[NSNull null]]) {
    database = [FIRDatabase databaseWithURL:databaseURL];
  } else {
    database = [FIRDatabase database];
  }
  void (^defaultCompletionBlock)(NSError *, FIRDatabaseReference *) =
      ^(NSError *error, FIRDatabaseReference *ref) {
        result(getFlutterError(error));
      };
  if ([@"FirebaseDatabase#goOnline" isEqualToString:call.method]) {
    [database goOnline];
    result(nil);
  } else if ([@"FirebaseDatabase#goOffline" isEqualToString:call.method]) {
    [database goOffline];
    result(nil);
  } else if ([@"FirebaseDatabase#purgeOutstandingWrites" isEqualToString:call.method]) {
    [database purgeOutstandingWrites];
    result(nil);
  } else if ([@"FirebaseDatabase#setPersistenceEnabled" isEqualToString:call.method]) {
    NSNumber *value = call.arguments[@"enabled"];
    @try {
      database.persistenceEnabled = value.boolValue;
      result([NSNumber numberWithBool:YES]);
    } @catch (NSException *exception) {
      if ([@"FIRDatabaseAlreadyInUse" isEqualToString:exception.name]) {
        // Database is already in use, e.g. after hot reload/restart.
        result([NSNumber numberWithBool:NO]);
      } else {
        @throw;
      }
    }
  } else if ([@"FirebaseDatabase#setPersistenceCacheSizeBytes" isEqualToString:call.method]) {
    NSNumber *value = call.arguments[@"cacheSize"];
    @try {
      database.persistenceCacheSizeBytes = value.unsignedIntegerValue;
      result([NSNumber numberWithBool:YES]);
    } @catch (NSException *exception) {
      if ([@"FIRDatabaseAlreadyInUse" isEqualToString:exception.name]) {
        // Database is already in use, e.g. after hot reload/restart.
        result([NSNumber numberWithBool:NO]);
      } else {
        @throw;
      }
    }
  } else if ([@"DatabaseReference#set" isEqualToString:call.method]) {
    [getReference(database, call.arguments) setValue:call.arguments[@"value"]
                                         andPriority:call.arguments[@"priority"]
                                 withCompletionBlock:defaultCompletionBlock];
  } else if ([@"DatabaseReference#update" isEqualToString:call.method]) {
    [getReference(database, call.arguments) updateChildValues:call.arguments[@"value"]
                                          withCompletionBlock:defaultCompletionBlock];
  } else if ([@"DatabaseReference#setPriority" isEqualToString:call.method]) {
    [getReference(database, call.arguments) setPriority:call.arguments[@"priority"]
                                    withCompletionBlock:defaultCompletionBlock];
  } else if ([@"DatabaseReference#runTransaction" isEqualToString:call.method]) {
    [getReference(database, call.arguments)
        runTransactionBlock:^FIRTransactionResult *_Nonnull(FIRMutableData *_Nonnull currentData) {
          // Create semaphore to allow native side to wait while snapshot
          // updates occur on the Dart side.
          dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);

          NSObject *snapshot =
              @{@"key" : currentData.key ?: [NSNull null], @"value" : currentData.value};

          __block bool shouldAbort = false;

          [self.channel invokeMethod:@"DoTransaction"
                           arguments:@{
                             @"transactionKey" : call.arguments[@"transactionKey"],
                             @"snapshot" : snapshot
                           }
                              result:^(id _Nullable result) {
                                if ([result isKindOfClass:[FlutterError class]]) {
                                  FlutterError *flutterError = ((FlutterError *)result);
                                  NSLog(@"Error code: %@", flutterError.code);
                                  NSLog(@"Error message: %@", flutterError.message);
                                  NSLog(@"Error details: %@", flutterError.details);
                                  shouldAbort = true;
                                } else if ([result isEqual:FlutterMethodNotImplemented]) {
                                  NSLog(@"DoTransaction not implemented on the Dart side.");
                                  shouldAbort = true;
                                } else {
                                  [self.updatedSnapshots
                                      setObject:result
                                         forKey:call.arguments[@"transactionKey"]];
                                }
                                dispatch_semaphore_signal(semaphore);
                              }];

          // Wait while Dart side updates the snapshot. Incoming transactionTimeout is in
          // milliseconds so converting to nanoseconds for use with dispatch_semaphore_wait.
          long result = dispatch_semaphore_wait(
              semaphore,
              dispatch_time(DISPATCH_TIME_NOW,
                            [call.arguments[@"transactionTimeout"] integerValue] * 1000000));

          if (result == 0 && !shouldAbort) {
            // Set FIRMutableData value to value returned from the Dart side.
            currentData.value =
                [self.updatedSnapshots objectForKey:call.arguments[@"transactionKey"]][@"value"];
          } else {
            if (result != 0) {
              NSLog(@"Transaction at %@ timed out.", [getReference(database, call.arguments) URL]);
            }
            return [FIRTransactionResult abort];
          }

          return [FIRTransactionResult successWithValue:currentData];
        }
        andCompletionBlock:^(NSError *_Nullable error, BOOL committed,
                             FIRDataSnapshot *_Nullable snapshot) {
          // Invoke transaction completion on the Dart side.
          result(@{
            @"transactionKey" : call.arguments[@"transactionKey"],
            @"error" : getDictionaryFromError(error) ?: [NSNull null],
            @"committed" : [NSNumber numberWithBool:committed],
            @"snapshot" : @{@"key" : snapshot.key ?: [NSNull null], @"value" : snapshot.value}
          });
        }];
  } else if ([@"OnDisconnect#set" isEqualToString:call.method]) {
    [getReference(database, call.arguments) onDisconnectSetValue:call.arguments[@"value"]
                                                     andPriority:call.arguments[@"priority"]
                                             withCompletionBlock:defaultCompletionBlock];
  } else if ([@"OnDisconnect#update" isEqualToString:call.method]) {
    [getReference(database, call.arguments) onDisconnectUpdateChildValues:call.arguments[@"value"]
                                                      withCompletionBlock:defaultCompletionBlock];
  } else if ([@"OnDisconnect#cancel" isEqualToString:call.method]) {
    [getReference(database, call.arguments)
        cancelDisconnectOperationsWithCompletionBlock:defaultCompletionBlock];
  } else if ([@"Query#observe" isEqualToString:call.method]) {
    FIRDataEventType eventType = parseEventType(call.arguments[@"eventType"]);
    __block FIRDatabaseHandle handle = [getDatabaseQuery(database, call.arguments)
        observeEventType:eventType
        andPreviousSiblingKeyWithBlock:^(FIRDataSnapshot *snapshot, NSString *previousSiblingKey) {
          [self.channel invokeMethod:@"Event"
                           arguments:@{
                             @"handle" : [NSNumber numberWithUnsignedInteger:handle],
                             @"snapshot" : @{
                               @"key" : snapshot.key ?: [NSNull null],
                               @"value" : roundDoubles(snapshot.value) ?: [NSNull null],
                             },
                             @"previousSiblingKey" : previousSiblingKey ?: [NSNull null],
                           }];
        }
        withCancelBlock:^(NSError *error) {
          [self.channel invokeMethod:@"Error"
                           arguments:@{
                             @"handle" : [NSNumber numberWithUnsignedInteger:handle],
                             @"error" : getDictionaryFromError(error),
                           }];
        }];
    result([NSNumber numberWithUnsignedInteger:handle]);
  } else if ([@"Query#removeObserver" isEqualToString:call.method]) {
    FIRDatabaseHandle handle = [call.arguments[@"handle"] unsignedIntegerValue];
    [getDatabaseQuery(database, call.arguments) removeObserverWithHandle:handle];
    result(nil);
  } else if ([@"Query#keepSynced" isEqualToString:call.method]) {
    NSNumber *value = call.arguments[@"value"];
    [getDatabaseQuery(database, call.arguments) keepSynced:value.boolValue];
    result(nil);
  } else {
    result(FlutterMethodNotImplemented);
  }
}

@end

File: packages/firebase_database/android/src/main/java/io/flutter/plugins/firebase/database/FirebaseDatabasePlugin.java
// Copyright 2017 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.firebase.database;

import android.util.Log;
import android.util.SparseArray;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.TaskCompletionSource;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.FirebaseApp;
import com.google.firebase.database.ChildEventListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseException;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.MutableData;
import com.google.firebase.database.Query;
import com.google.firebase.database.Transaction;
import com.google.firebase.database.ValueEventListener;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

/** FirebaseDatabasePlugin */
public class FirebaseDatabasePlugin implements MethodCallHandler {

  private static final String TAG = "FirebaseDatabasePlugin";

  private final MethodChannel channel;
  private static final String EVENT_TYPE_CHILD_ADDED = "_EventType.childAdded";
  private static final String EVENT_TYPE_CHILD_REMOVED = "_EventType.childRemoved";
  private static final String EVENT_TYPE_CHILD_CHANGED = "_EventType.childChanged";
  private static final String EVENT_TYPE_CHILD_MOVED = "_EventType.childMoved";
  private static final String EVENT_TYPE_VALUE = "_EventType.value";

  // Handles are ints used as indexes into the sparse array of active observers
  private int nextHandle = 0;
  private final SparseArray<EventObserver> observers = new SparseArray<>();

  public static void registerWith(PluginRegistry.Registrar registrar) {
    final MethodChannel channel =
        new MethodChannel(registrar.messenger(), "plugins.flutter.io/firebase_database");
    channel.setMethodCallHandler(new FirebaseDatabasePlugin(channel));
  }

  private FirebaseDatabasePlugin(MethodChannel channel) {
    this.channel = channel;
  }

  private DatabaseReference getReference(FirebaseDatabase database, Map<String, Object> arguments) {
    String path = (String) arguments.get("path");
    DatabaseReference reference = database.getReference();
    if (path != null) reference = reference.child(path);
    return reference;
  }

  private Query getQuery(FirebaseDatabase database, Map<String, Object> arguments) {
    Query query = getReference(database, arguments);
    @SuppressWarnings("unchecked")
    Map<String, Object> parameters = (Map<String, Object>) arguments.get("parameters");
    if (parameters == null) return query;
    Object orderBy = parameters.get("orderBy");
    if ("child".equals(orderBy)) {
      query = query.orderByChild((String) parameters.get("orderByChildKey"));
    } else if ("key".equals(orderBy)) {
      query = query.orderByKey();
    } else if ("value".equals(orderBy)) {
      query = query.orderByValue();
    } else if ("priority".equals(orderBy)) {
      query = query.orderByPriority();
    }
    if (parameters.containsKey("startAt")) {
      Object startAt = parameters.get("startAt");
      if (parameters.containsKey("startAtKey")) {
        String startAtKey = (String) parameters.get("startAtKey");
        if (startAt instanceof Boolean) {
          query = query.startAt((Boolean) startAt, startAtKey);
        } else if (startAt instanceof Number) {
          query = query.startAt(((Number) startAt).doubleValue(), startAtKey);
        } else {
          query = query.startAt((String) startAt, startAtKey);
        }
      } else {
        if (startAt instanceof Boolean) {
          query = query.startAt((Boolean) startAt);
        } else if (startAt instanceof Number) {
          query = query.startAt(((Number) startAt).doubleValue());
        } else {
          query = query.startAt((String) startAt);
        }
      }
    }
    if (parameters.containsKey("endAt")) {
      Object endAt = parameters.get("endAt");
      if (parameters.containsKey("endAtKey")) {
        String endAtKey = (String) parameters.get("endAtKey");
        if (endAt instanceof Boolean) {
          query = query.endAt((Boolean) endAt, endAtKey);
        } else if (endAt instanceof Number) {
          query = query.endAt(((Number) endAt).doubleValue(), endAtKey);
        } else {
          query = query.endAt((String) endAt, endAtKey);
        }
      } else {
        if (endAt instanceof Boolean) {
          query = query.endAt((Boolean) endAt);
        } else if (endAt instanceof Number) {
          query = query.endAt(((Number) endAt).doubleValue());
        } else {
          query = query.endAt((String) endAt);
        }
      }
    }
    if (parameters.containsKey("equalTo")) {
      Object equalTo = parameters.get("equalTo");
      if (parameters.containsKey("equalToKey")) {
        String equalToKey = (String) parameters.get("equalToKey");
        if (equalTo instanceof Boolean) {
          query = query.equalTo((Boolean) equalTo, equalToKey);
        } else if (equalTo instanceof Number) {
          query = query.equalTo(((Number) equalTo).doubleValue(), equalToKey);
        } else {
          query = query.equalTo((String) equalTo, equalToKey);
        }
      } else {
        if (equalTo instanceof Boolean) {
          query = query.equalTo((Boolean) equalTo);
        } else if (equalTo instanceof Number) {
          query = query.equalTo(((Number) equalTo).doubleValue());
        } else {
          query = query.equalTo((String) equalTo);
        }
      }
    }
    if (parameters.containsKey("limitToFirst")) {
      query = query.limitToFirst((int) parameters.get("limitToFirst"));
    }
    if (parameters.containsKey("limitToLast")) {
      query = query.limitToLast((int) parameters.get("limitToLast"));
    }
    return query;
  }

  private class DefaultCompletionListener implements DatabaseReference.CompletionListener {
    private final Result result;

    DefaultCompletionListener(Result result) {
      this.result = result;
    }

    @Override
    public void onComplete(DatabaseError error, DatabaseReference ref) {
      if (error != null) {
        result.error(String.valueOf(error.getCode()), error.getMessage(), error.getDetails());
      } else {
        result.success(null);
      }
    }
  }

  private class EventObserver implements ChildEventListener, ValueEventListener {
    private String requestedEventType;
    private int handle;

    EventObserver(String requestedEventType, int handle) {
      this.requestedEventType = requestedEventType;
      this.handle = handle;
    }

    private void sendEvent(String eventType, DataSnapshot snapshot, String previousChildName) {
      if (eventType.equals(requestedEventType)) {
        Map<String, Object> arguments = new HashMap<>();
        Map<String, Object> snapshotMap = new HashMap<>();
        snapshotMap.put("key", snapshot.getKey());
        snapshotMap.put("value", snapshot.getValue());
        arguments.put("handle", handle);
        arguments.put("snapshot", snapshotMap);
        arguments.put("previousSiblingKey", previousChildName);
        channel.invokeMethod("Event", arguments);
      }
    }

    @Override
    public void onCancelled(DatabaseError error) {
      Map<String, Object> arguments = new HashMap<>();
      arguments.put("handle", handle);
      arguments.put("error", asMap(error));
      channel.invokeMethod("Error", arguments);
    }

    @Override
    public void onChildAdded(DataSnapshot snapshot, String previousChildName) {
      sendEvent(EVENT_TYPE_CHILD_ADDED, snapshot, previousChildName);
    }

    @Override
    public void onChildRemoved(DataSnapshot snapshot) {
      sendEvent(EVENT_TYPE_CHILD_REMOVED, snapshot, null);
    }

    @Override
    public void onChildChanged(DataSnapshot snapshot, String previousChildName) {
      sendEvent(EVENT_TYPE_CHILD_CHANGED, snapshot, previousChildName);
    }

    @Override
    public void onChildMoved(DataSnapshot snapshot, String previousChildName) {
      sendEvent(EVENT_TYPE_CHILD_MOVED, snapshot, previousChildName);
    }

    @Override
    public void onDataChange(DataSnapshot snapshot) {
      sendEvent(EVENT_TYPE_VALUE, snapshot, null);
    }
  }

  @Override
  public void onMethodCall(final MethodCall call, final Result result) {
    final Map<String, Object> arguments = call.arguments();
    FirebaseDatabase database;
    String appName = (String) arguments.get("app");
    String databaseURL = (String) arguments.get("databaseURL");
    if (appName != null && databaseURL != null) {
      database = FirebaseDatabase.getInstance(FirebaseApp.getInstance(appName), databaseURL);
    } else if (appName != null) {
      database = FirebaseDatabase.getInstance(FirebaseApp.getInstance(appName));
    } else if (databaseURL != null) {
      database = FirebaseDatabase.getInstance(databaseURL);
    } else {
      database = FirebaseDatabase.getInstance();
    }
    switch (call.method) {
      case "FirebaseDatabase#goOnline":
        {
          database.goOnline();
          result.success(null);
          break;
        }

      case "FirebaseDatabase#goOffline":
        {
          database.goOffline();
          result.success(null);
          break;
        }

      case "FirebaseDatabase#purgeOutstandingWrites":
        {
          database.purgeOutstandingWrites();
          result.success(null);
          break;
        }

      case "FirebaseDatabase#setPersistenceEnabled":
        {
          Boolean isEnabled = (Boolean) arguments.get("enabled");
          try {
            database.setPersistenceEnabled(isEnabled);
            result.success(true);
          } catch (DatabaseException e) {
            // Database is already in use, e.g. after hot reload/restart.
            result.success(false);
          }
          break;
        }

      case "FirebaseDatabase#setPersistenceCacheSizeBytes":
        {
          long cacheSize = (Integer) arguments.get("cacheSize");
          try {
            database.setPersistenceCacheSizeBytes(cacheSize);
            result.success(true);
          } catch (DatabaseException e) {
            // Database is already in use, e.g. after hot reload/restart.
            result.success(false);
          }
          break;
        }

      case "DatabaseReference#set":
        {
          Object value = arguments.get("value");
          Object priority = arguments.get("priority");
          DatabaseReference reference = getReference(database, arguments);
          if (priority != null) {
            reference.setValue(value, priority, new DefaultCompletionListener(result));
          } else {
            reference.setValue(value, new DefaultCompletionListener(result));
          }
          break;
        }

      case "DatabaseReference#update":
        {
          @SuppressWarnings("unchecked")
          Map<String, Object> value = (Map<String, Object>) arguments.get("value");
          DatabaseReference reference = getReference(database, arguments);
          reference.updateChildren(value, new DefaultCompletionListener(result));
          break;
        }

      case "DatabaseReference#setPriority":
        {
          Object priority = arguments.get("priority");
          DatabaseReference reference = getReference(database, arguments);
          reference.setPriority(priority, new DefaultCompletionListener(result));
          break;
        }

      case "DatabaseReference#runTransaction":
        {
          final DatabaseReference reference = getReference(database, arguments);

          // Initiate native transaction.
          reference.runTransaction(
              new Transaction.Handler() {
                @Override
                public Transaction.Result doTransaction(MutableData mutableData) {
                  // Tasks are used to allow native execution of doTransaction to wait while Snapshot is
                  // processed by logic on the Dart side.
                  final TaskCompletionSource<Map<String, Object>> updateMutableDataTCS =
                      new TaskCompletionSource<>();
                  final Task<Map<String, Object>> updateMutableDataTCSTask =
                      updateMutableDataTCS.getTask();

                  Map<String, Object> doTransactionMap = new HashMap<>();
                  doTransactionMap.put("transactionKey", arguments.get("transactionKey"));

                  final Map<String, Object> snapshotMap = new HashMap<>();
                  snapshotMap.put("key", mutableData.getKey());
                  snapshotMap.put("value", mutableData.getValue());
                  doTransactionMap.put("snapshot", snapshotMap);

                  // Return snapshot to Dart side for update.
                  channel.invokeMethod(
                      "DoTransaction",
                      doTransactionMap,
                      new MethodChannel.Result() {
                        @Override
                        @SuppressWarnings("unchecked")
                        public void success(Object result) {
                          updateMutableDataTCS.setResult((Map<String, Object>) result);
                        }

                        @Override
                        public void error(
                            String errorCode, String errorMessage, Object errorDetails) {
                          String exceptionMessage =
                              "Error code: "
                                  + errorCode
                                  + "\nError message: "
                                  + errorMessage
                                  + "\nError details: "
                                  + errorDetails;
                          updateMutableDataTCS.setException(new Exception(exceptionMessage));
                        }

                        @Override
                        public void notImplemented() {
                          updateMutableDataTCS.setException(
                              new Exception("DoTransaction not implemented on Dart side."));
                        }
                      });

                  try {
                    // Wait for updated snapshot from the Dart side.
                    Map<String, Object> updatedSnapshotMap =
                        Tasks.await(
                            updateMutableDataTCSTask,
                            (int) arguments.get("transactionTimeout"),
                            TimeUnit.MILLISECONDS);
                    // Set value of MutableData to value returned from the Dart side.
                    mutableData.setValue(updatedSnapshotMap.get("value"));
                  } catch (ExecutionException | InterruptedException | TimeoutException e) {
                    Log.e(TAG, "Unable to commit Snapshot update. Transaction failed.", e);
                    if (e instanceof TimeoutException) {
                      Log.e(TAG, "Transaction at " + reference.toString() + " timed out.");
                    }
                    return Transaction.abort();
                  }
                  return Transaction.success(mutableData);
                }

                @Override
                public void onComplete(
                    DatabaseError databaseError, boolean committed, DataSnapshot dataSnapshot) {
                  Map<String, Object> completionMap = new HashMap<>();
                  completionMap.put("transactionKey", arguments.get("transactionKey"));
                  if (databaseError != null) {
                    completionMap.put("error", asMap(databaseError));
                  }
                  completionMap.put("committed", committed);
                  if (dataSnapshot != null) {
                    Map<String, Object> snapshotMap = new HashMap<>();
                    snapshotMap.put("key", dataSnapshot.getKey());
                    snapshotMap.put("value", dataSnapshot.getValue());
                    completionMap.put("snapshot", snapshotMap);
                  }

                  // Invoke transaction completion on the Dart side.
                  result.success(completionMap);
                }
              });
          break;
        }

      case "OnDisconnect#set":
        {
          Object value = arguments.get("value");
          Object priority = arguments.get("priority");
          DatabaseReference reference = getReference(database, arguments);
          if (priority != null) {
            if (priority instanceof String) {
              reference
                  .onDisconnect()
                  .setValue(value, (String) priority, new DefaultCompletionListener(result));
            } else if (priority instanceof Double) {
              reference
                  .onDisconnect()
                  .setValue(value, (double) priority, new DefaultCompletionListener(result));
            } else if (priority instanceof Map) {
              reference
                  .onDisconnect()
                  .setValue(value, (Map) priority, new DefaultCompletionListener(result));
            }
          } else {
            reference.onDisconnect().setValue(value, new DefaultCompletionListener(result));
          }
          break;
        }

      case "OnDisconnect#update":
        {
          @SuppressWarnings("unchecked")
          Map<String, Object> value = (Map<String, Object>) arguments.get("value");
          DatabaseReference reference = getReference(database, arguments);
          reference.onDisconnect().updateChildren(value, new DefaultCompletionListener(result));
          break;
        }

      case "OnDisconnect#cancel":
        {
          DatabaseReference reference = getReference(database, arguments);
          reference.onDisconnect().cancel(new DefaultCompletionListener(result));
          break;
        }

      case "Query#keepSynced":
        {
          boolean value = (Boolean) arguments.get("value");
          getQuery(database, arguments).keepSynced(value);
          result.success(null);
          break;
        }

      case "Query#observe":
        {
          String eventType = (String) arguments.get("eventType");
          int handle = nextHandle++;
          EventObserver observer = new EventObserver(eventType, handle);
          observers.put(handle, observer);
          if (eventType.equals(EVENT_TYPE_VALUE)) {
            getQuery(database, arguments).addValueEventListener(observer);
          } else {
            getQuery(database, arguments).addChildEventListener(observer);
          }
          result.success(handle);
          break;
        }

      case "Query#removeObserver":
        {
          Query query = getQuery(database, arguments);
          int handle = (Integer) arguments.get("handle");
          EventObserver observer = observers.get(handle);
          if (observer != null) {
            if (observer.requestedEventType.equals(EVENT_TYPE_VALUE)) {
              query.removeEventListener((ValueEventListener) observer);
            } else {
              query.removeEventListener((ChildEventListener) observer);
            }
            observers.delete(handle);
            result.success(null);
            break;
          } else {
            result.error("unknown_handle", "removeObserver called on an unknown handle", null);
            break;
          }
        }

      default:
        {
          result.notImplemented();
          break;
        }
    }
  }

  private static Map<String, Object> asMap(DatabaseError error) {
    Map<String, Object> map = new HashMap<>();
    map.put("code", error.getCode());
    map.put("message", error.getMessage());
    map.put("details", error.getDetails());
    return map;
  }
}
